mqt/qmap/subarchitectures.py

```diff
@@ -0,0 +1,275 @@
+
+
+class SubarchitectureOrder:
+    """
+    Class representing partial order of Subarchitectures.
+
+    Attributes
+    ----------
+    arch : Architecture
+        quantum computing architecture whose subarchitectures should be ordered
+    sgs: list[list[Architecture]]
+        subarchitectures of arch. sgs[i][j]: the j-th subarchitecture of size i
+    subarch_order : dict[tuple[int, int], list[tuple[int, int]]]
+        ordering of variables according to subarchitecture order.
+        tuples (n, i) are indices into self.sgs
+    desirable_subarchitectures: dict[tuple[int, int], list[tuple[int, int]]]
+        mapping of indices (n, i) to its desirable subarchitectures
+        tuples (n, i) are indices into self.sgs
+
+    Methods
+    -------
+    optimal_candidates(nqubits)
+        return optimal candidate for mapping quantum circuits of a given size
+    covering(nqubits, size)
+        return a covering for nqubit circuits limited by size
+    store_library(lib_name)
+        serialize this object to avoid recomputing the ordering in the future
+    """
+
+    def __init__(self, arch: Subarchitecture | str):
+        """
+        Initialize the partial order.
+
+        If an architecture is given, the order will be computed for this
+        specific architecture.
+        If a str is given instead, the ordering will be loaded from the
+        subarchitecture library of that name.
+        """
+        if type(arch) is str:
+            self.__load_library(arch)
+            return
+
+        if isinstance(arch, rx.PyGraph):
+            self.arch = arch
+        else:
+            self.arch = rx.networkx_converter(nx.from_edgelist(arch))
+
+        self.subarch_order = defaultdict(lambda: [])
+        self.desirable_subarchitectures = defaultdict(lambda: [])
+
+        self.__isomorphisms = defaultdict(lambda: {})
+
+        self.__compute_subgraphs()
+        self.__compute_subarch_order()
+        self.__compute_desirable_subarchitectures()
+        self.subarch_order = dict(self.subarch_order)
+        self.desirable_subarchitectures = dict(self.desirable_subarchitectures)
+        self.__isomorphisms = dict(self.__isomorphisms)
+
+    def optimal_candidates(self, nqubits: int) -> list[Subarchitecture]:
+        """Return optimal subarchitecture candidate."""
+        if nqubits <= 0 or nqubits > self.arch.num_nodes():
+            raise ValueError(
+                "Number of qubits must not be smaller or equal 0 or larger then number of physical qubits of architecture."
+            )
+
+        if nqubits == self.arch.num_nodes():
+            return [self.arch]
+
+        cands = self.__cand(nqubits)
+        trans_ord = self.__transitive_closure(self.subarch_order)
+        ref_ord = self.__reflexive_closure(trans_ord)
+
+        opt_cands = set(ref_ord[next(iter(cands))])
+        for cand in cands:
+            opt_cands = opt_cands.intersection(set(ref_ord[cand]))
+
+        ordered_cands = list(opt_cands)
```

Comment:
Does incur a copy, but that is desired. opt_cands is transformed in a loop that iterates over the elements of opt_cands. Since the iterator shouldn't be invalidated, the copy is needed.